File: Chapter13/Java/TestHandlers/app/src/main/java/com/androidbook/handlers/WorkerThreadRunnable.java
package com.androidbook.handlers;

import android.os.Handler;
import android.os.Message;
import android.util.Log;

public class WorkerThreadRunnable implements Runnable {

    private static final String TAG = "TestRunnable";

    private final Handler mainThreadHandler;

    public WorkerThreadRunnable(Handler h) {
        mainThreadHandler = h;
    }

    @Override
    public void run() {
        Log.d(TAG, "start execution");
        informStart();
        for (int i = 1; i <= 10; i++) {
            Utils.sleepForInSecs(1);
            informMiddle(i);
        }
        informFinish();
    }

    private void informMiddle(int count) {
        Message m = mainThreadHandler.obtainMessage();
        m.setData(Utils.getStringAsABundle("done:" + count));
        mainThreadHandler.sendMessage(m);
    }

    private void informStart() {
        Message m = mainThreadHandler.obtainMessage();
        m.setData(Utils.getStringAsABundle("starting run"));
        mainThreadHandler.sendMessage(m);
    }

    private void informFinish() {
        Message m = mainThreadHandler.obtainMessage();
        m.setData(Utils.getStringAsABundle("Finishing run"));
        mainThreadHandler.sendMessage(m);
    }

}

File: Chapter13/Kotlin/TestHandlers/app/src/main/java/com/androidbook/handlers/WorkerThreadRunnable.kt
package com.androidbook.handlers

import android.os.Handler
import android.util.Log

class WorkerThreadRunnable(private val mainThreadHandler: Handler) : Runnable {

    override fun run() {
        Log.d(TAG, "start execution")
        informStart()
        for (i in 1..10) {
            Utils.sleepForInSecs(1)
            informMiddle(i)
        }
        informFinish()
    }

    private fun informMiddle(count: Int) {
        val m = mainThreadHandler.obtainMessage()
        m.data = Utils.getStringAsABundle("done:$count")
        mainThreadHandler.sendMessage(m)
    }

    private fun informStart() {
        val m = mainThreadHandler.obtainMessage()
        m.data = Utils.getStringAsABundle("starting run")
        mainThreadHandler.sendMessage(m)
    }

    private fun informFinish() {
        val m = mainThreadHandler.obtainMessage()
        m.data = Utils.getStringAsABundle("Finishing run")
        mainThreadHandler.sendMessage(m)
    }

    companion object {
        const val TAG = "TestRunnable"
    }
}